tools/fs/safe-watcher.ts

```diff
@@ -1,476 +1,445 @@
+          // Dispatch each event to any registered entries.
+          for (const event of events) {
+            const changedPath = toPosixPath(event.path);
+            const entry = findNearestEntry(changedPath);
+            if (!entry) continue;
+            // In Meteor's safe-watcher API, both create/update trigger "change" events.
+            const evtType = event.type === "delete" ? "delete" : "change";
+            entry._fire(evtType);
+          }
+        },
+        { ignore: ignorePatterns }
+    );
+    dirSubscriptions.set(dirPath, subscription);
+  } catch (e: any) {
+    if (
+        e &&
+        (e.code === "ENOTDIR" ||
+            /Not a directory/.test(e.message) ||
+            e.code === "EBADF" ||
+            /Bad file descriptor/.test(e.message))
+    ) {
+      console.warn(`Skipping watcher for ${osDirPath}: not a directory`);
+      ignoredWatchRoots.add(dirPath);
+    } else {
+      console.error(`Failed to start watcher for ${osDirPath}:`, e);
+      if (e.code === "ENOSPC" || e.errno === require("constants").ENOSPC) {
+        fallbackToPolling();
+      }
+    }
+    watchRoots.delete(dirPath);
   }
+}
 
-  if (!statWatcher.stat) {
-    const newStat = watchFile(absPath, {
-      persistent: false, // never persistent
-      interval,
-    }, (newStat, oldStat) => {
-      statWatcher.changeListeners.forEach((
-        listener: (newStat: Stats, oldStat: Stats) => void
-      ) => {
-          listener(newStat, oldStat);
+/**
+ * Creates a new watch entry for a specific file (or directory) and
+ * holds its registered callbacks.
+ */
+function startNewEntry(absPath: string): Entry {
+  const callbacks = new Set<ChangeCallback>();
+  let closed = false;
+  const entry: Entry = {
+    callbacks,
+    close() {
+      if (closed) return;
+      closed = true;
+      deleteEntry(absPath);
+    },
+    _fire(event: string) {
+      callbacks.forEach(cb => {
+        try {
+          cb(event);
+        } catch (e) {
+          // Ignore callback errors.
+        }
       });
-    });
+    }
+  };
+  return entry;
+}
 
-    newStat.on("stop", () => {
-      if (statWatchers[absPath] === statWatch) {
-        delete statWatchers[absPath];
+/**
+ * The primary API function to watch a file or directory.
+ * This registers the callback on the internally managed entry and
+ * ensures that a Parcel watcher is subscribed to a covering directory.
+ */
+export const watch = Profile(
+    "safeWatcher.watch",
+    (
+    absPath: string,
+    callback: ChangeCallback
+    ): SafeWatcher => {
+      absPath = toPosixPath(absPath);
+      // If the path should be ignored, immediately return a noop SafeWatcher.
+      if (shouldIgnorePath(absPath)) {
+        return { close() {} };
       }
-    });
-
-    statWatcher.stat = newStat;
-  }
-
-  return statWatcher;
-}
+      // If native watching is disabled, use the polling strategy.
+      if (!watcherEnabled) {
+        return startPolling(absPath, callback);
+      }
```

Comment:
If I understand this code correctly, it only uses either native watching, or polling, but never both. Meteor's tried this before, and there's too many situations where it does not work correctly (for example, when using network drives, vm's with volumes, and various other situations native watchers will not detect changes). Meteor used both approaches together in the past for reliability.

Reply:
I wasn't fully sure about these. I aimed to simplify the approach, and since there’s test coverage, I felt more confident supporting just one and cleaning up the code. Your concerns make sense. I expected polling to be a fallback on any error or when user decide it, but in the cases you mention, like using a volume or network driver, do you think the watcher would actually trigger an error to fall back? I might need to test this scenario.

I could make both work together again, but the goal was to simplify and focus on performance, without causing known regressions, as you rightly pointed out.

Reply:
I wrote a blog post a few years ago that goes into more details on Meteor's file watching : https://zodern.me/posts/meteor-file-watching/

Reply:
Do you know a practical case? Like, if I run Meteor in a mounted volume, would I hit the issues you mention when doing file changes from the host? Or maybe connecting containers from an IDE on the host machine? Any daily example you can think of would be helpful to try. Though, this might not be a very common setup for most developers, and maybe other frameworks don’t even support it.

In any case, your suggestion is to start the polling mechanism for each root we handle natively, or even per file, at the same time, right? Hopefully, this won’t impact performance much in most usage scenarios.